File: Xsquare And Chocolates Bars/Xsquare And Chocolates Bars.py
T = input()

for _ in xrange(T):
    string = raw_input().strip()
    length = len(string)
    DP = [0] * (length + 1)
    
    if length <= 2:
        print length
    
    if length == 3:
        if string[0] == string[1] and string[0] == string[2]:
            print length
            continue
        print 0
        continue
    DP[0] = 0
    DP[1] = 0
    DP[2] = 0
    Max = 0
    for i in xrange( 3, length + 1 ):
        if string[i - 1] == string[i - 2] and string[i - 1] == string[i - 3]:
            DP[i] = 0
        else:
            DP[i] = DP[i - 3] + 3
        if DP[i] > Max:
            Max = DP[i]
    print length - Max
    

File: Xsquare And Chocolates Bars/Xsquare And Chocolates Bars.cpp
#include <iostream>
#include <vector>
using namespace std;

int main()
{
    int N = 0;
    string str = "";
    
    cin>>N;
    for( int i = 0; i < N; i++ )
    {        
        
        cin>>str;
        vector<int> DP(str.length() + 1, 0);
        if( str.length() <= 2 )
        {
            cout<<str.length()<<endl;
            continue;
        }
        
        if( str.length() <= 3 )
        {
            if( str[0] == str[1] && str[0] == str[2] )
            {
                cout<<str.length()<<endl;
                continue;
            }
            cout<<0<<endl;
            continue;
        }
        
        int Max = 0;
        for( int j = 3; j < str.length() + 1; j++ )
        {
            if( str[j - 1] == str[j - 2] && str[j - 1] == str[j - 3] )
            {
                DP[j] = 0;
            }
            else
            {
                DP[j] = DP[j - 3] + 3;
            }
            Max = max(DP[j], Max);
        }
        
        cout<<str.length() - Max<<endl;
    }
    
}
